### PR TITLE
Support Group By _wavefront_source Dimension and Spy on Id Creation

### PR DESCRIPTION
**Pull request by ktjoanna**
Changes regarding issue #3.

### CODE REVIEW

File: src/main/java/com/wavefront/tools/wftop/components/RootNode.java [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+     * @param groupByIngestionSource Choose whether to group by Ingestion Source.
+     * @param accessed              % Accessed.
+     * @param metric                Metric name.
+     * @param host                  Host name.
+     * @param pointTags             Point Tags map.
+     * @param timestamp             Time of metric ingestion.
+     * @param value                 Point value.
+     */
+    public void accept(Dimension analysisDimension, boolean groupByIngestionSource,
+           boolean accessed, String metric, String host, Multimap<String, String> pointTags, long timestamp, double value){
```
Comment (panghy) [inline src/main/java/com/wavefront/tools/wftop/components/RootNode.java]: Align paramsters and don’t have lines > 100

File: src/main/java/com/wavefront/tools/wftop/components/RootNode.java [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+        String source_value = "None";
+        if (groupByIngestionSource) {
+            for (Map.Entry<String, String> s : pointTags.entries()){
+                if (s.getKey().equals(WAVEFRONT_SOURCE_TAG_KEY)){
+                    source_value = pointTags.get(WAVEFRONT_SOURCE_TAG_KEY).iterator().next();
+                    break;
+                }
+            }
+            if (ingestionSource.get(source_value) == null){
```
Comment (panghy) [inline src/main/java/com/wavefront/tools/wftop/components/RootNode.java]: space before {

File: src/main/java/com/wavefront/tools/wftop/components/RootNode.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+                ingestionSource.put(source_value, new SourceNode(source_value));
+                ingestionSource.get(source_value).getNamespaceBuilder().setSeparatorCharacters(getSeparatorCharacters());
+                ingestionSource.get(source_value).getNamespaceBuilder().setMaxDepth(getMaxDepth());
+                ingestionSource.get(source_value).getNamespaceBuilder().setMaxChildren(getMaxChildren());
+            }
+        }
+        ingestionSource.get(source_value).accept(analysisDimension, accessed, metric, host,
+                pointTags, timestamp, value);
```
Comment (panghy) [inline src/main/java/com/wavefront/tools/wftop/components/RootNode.java]: this looks too far indented check your style settings

File: src/main/java/com/wavefront/tools/wftop/components/RootNode.java [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        globalNode.accept(analysisDimension, accessed, metric, host,
+                pointTags, timestamp, value);
```
Comment (panghy) [inline src/main/java/com/wavefront/tools/wftop/components/RootNode.java]: same here

File: src/main/java/com/wavefront/tools/wftop/components/RootNode.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+    public int getMaxChildren() {
+        return branchLimit;
+    }
+
+    /**
+     * @return SourceNode "None", the default when Group By Source not specified.
+     */
+    public SourceNode getDefaultRoot(){
+        for (Map.Entry<String, SourceNode> entry : ingestionSource.entrySet()){
```
Comment (panghy) [inline src/main/java/com/wavefront/tools/wftop/components/RootNode.java]: actually all your indents are too far (check your intellij settings)

File: src/main/java/com/wavefront/tools/wftop/components/SourceNode.java [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+import com.codahale.metrics.Meter;
+import com.google.common.util.concurrent.AtomicDouble;
+
+import java.util.Map;
+
+/**
+ * Groups metrics based on source of ingestion. Default source is set to "None".
+ *
+ * @author Joanna Ko (kjoanna@vmware.com).
+ */
```
Comment (panghy) [inline src/main/java/com/wavefront/tools/wftop/components/SourceNode.java:14]: same here, all your indents are too far

File: src/main/java/com/wavefront/tools/wftop/components/SourceNode.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+     * @param analysisDimension Analysis Dimension chosen.
+     * @param accessed          % Accessed.
+     * @param metric            Metric name.
+     * @param host              Host name.
+     * @param pointTags         Point Tags map.
+     * @param timestamp         Time of metric ingestion.
+     * @param value             Point value.
+     */
+    public void accept(Dimension analysisDimension, boolean accessed, String metric, String host,
+                          Multimap<String, String> pointTags, long timestamp, double value){
```
Comment (panghy) [inline src/main/java/com/wavefront/tools/wftop/components/SourceNode.java]: align params

**Review by panghy — COMMENTED** [review]
Check your style settings. 4/2 indent is what we use, you seem to be doing 8/4

**Comment by ktjoanna** [comment]
Adding spying on ID Creations for issue #2  on the same PR to avoid merge conflicts later on.

File: src/main/java/com/wavefront/tools/wftop/components/PointsSpy.java [hunk 8]
```diff
@@ -303,20 +358,69 @@ private void handleLine(String line) {
...
+          // ignored.
+        }
       }
+    } else {
+      String[] IdLine = line.split("\\s+");
+      if (isType(IdLine[0])) parseId(IdLine);
+    }
+  }
+
+  private boolean isType(String type) {
```
Comment (panghy) [inline src/main/java/com/wavefront/tools/wftop/components/PointsSpy.java]: Maybe spyable type?

File: src/main/java/com/wavefront/tools/wftop/components/PointsSpy.java [hunk 9]
```diff
@@ -329,15 +433,43 @@ private void parseMetric(boolean accessed, String line) {
...
         annotations = ImmutableMultimap.of();
       }
-      this.listener.onMetricReceived(this, accessed, drive.getMetric(), drive.getHost(), annotations,
-          drive.getTimestamp(), (Double) drive.getValue());
+      this.listener.onMetricReceived(this, accessed, drive.getMetric(), drive.getHost(),
+          annotations, drive.getTimestamp(), (Double) drive.getValue());
+    }
+  }
+
+  public void setTypePrefix(Type t) {
```
Comment (panghy) [inline src/main/java/com/wavefront/tools/wftop/components/PointsSpy.java:445]: public methods should have javadoc

**Review by panghy — COMMENTED** [review]
One more round, looks good otherwise

File: src/main/java/com/wavefront/tools/wftop/WavefrontTop.java [hunk 10]
```diff
@@ -181,33 +197,35 @@ public void onConnecting(PointsSpy pointsSpy) {
...
-        if (pointsSpy.isConnected()) {
-          refreshPointsNamespacePanel(samplingRate);
-        }
+      double samplingRate = pointsSpy.getSamplingRate();
+      namespacePanel.setGlobalPPS(Math.max(1, backendCount.get()) / samplingRate, root.getRate());
+      namespacePanel.setSamplingRate(samplingRate);
+      namespacePanel.setVisibleRows(gui.getScreen().getTerminalSize().getRows() - 10);
+      if (pointsSpy.isConnected()) {
+        refreshNamespacePanel(samplingRate);
+      }
```
Comment (panghy) [inline src/main/java/com/wavefront/tools/wftop/WavefrontTop.java]: indenting of this whole block is off, check "Reformat Code"

File: src/main/java/com/wavefront/tools/wftop/components/Node.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+   */
+  AtomicDouble getMax();
+
+  /**
+   * @return Range of values metric reports.
+   */
+  double getRange();
+
+  /**
+   * @return Map of current Node's children nodes
```
Comment (panghy) [inline src/main/java/com/wavefront/tools/wftop/components/Node.java]: period at end of sentence

File: src/main/java/com/wavefront/tools/wftop/components/Node.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+
+/**
+ * Interface for nodes, implemented by namespace and source nodes.
+ *
+ * @author Joanna Ko (kjoanna@vmware.com)
+ */
+public interface Node<T extends Node> {
+
+  /**
+   * @return Name of metric
```
Comment (panghy) [inline src/main/java/com/wavefront/tools/wftop/components/Node.java]: period at end of sentence

File: src/main/java/com/wavefront/tools/wftop/components/PointsSpy.java [hunk 13]
```diff
@@ -303,20 +358,69 @@ private void handleLine(String line) {
...
+          // ignored.
+        }
       }
+    } else {
+      String[] IdLine = line.split("\\s+");
+      if (spyableType(IdLine[0])) parseId(IdLine);
+    }
+  }
+
+  private boolean spyableType(String type) {
```
Comment (panghy) [inline src/main/java/com/wavefront/tools/wftop/components/PointsSpy.java]: isSpytableType

File: src/main/java/com/wavefront/tools/wftop/components/PointsSpy.java [hunk 14]
```diff
@@ -303,20 +358,69 @@ private void handleLine(String line) {
...
+      case "HISTOGRAM":
+      case "SPAN":
+      case "METRIC":
+        return true;
+      default:
+        return false;
+    }
+  }
+
+  private Type getType(String type) {
```
Comment (panghy) [inline src/main/java/com/wavefront/tools/wftop/components/PointsSpy.java]: toType

File: src/main/java/com/wavefront/tools/wftop/components/RootNode.java [hunk 15]
```diff
@@ -0,0 +1,192 @@
...
+    globalNode.accept(analysisDimension, accessed, metric, host,
+        pointTags, timestamp, value);
+  }
+
+  public String getSeparatorCharacters() {
+    return separators;
+  }
+
+  /**
+   * clears SourceNodes and adds default SourceNode "None".
```
Comment (panghy) [inline src/main/java/com/wavefront/tools/wftop/components/RootNode.java]: Clears

File: src/main/java/com/wavefront/tools/wftop/components/Type.java [hunk 16]
```diff
@@ -0,0 +1,13 @@
+package com.wavefront.tools.wftop.components;
+/**
```
Comment (panghy) [inline src/main/java/com/wavefront/tools/wftop/components/Type.java:3]: new line after package

File: src/main/java/com/wavefront/tools/wftop/panels/NamespacePanel.java [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+        columnLabel.substring(0, columnLabel.length() - 4));
+    columnLabel = this.table.getTableModel().getColumnLabel(sortIndex);
+    this.table.getTableModel().setColumnLabel(sortIndex, columnLabel + (reverseSort ? " [↑]" : " [↓]"));
+  }
+
+  public void setVisibleRows(int count) {
+    this.table.setVisibleRows(count);
+  }
+
+  protected abstract Comparator<Node> getComparator();
```
Comment (panghy) [inline src/main/java/com/wavefront/tools/wftop/panels/NamespacePanel.java]: javadoc

File: src/main/java/com/wavefront/tools/wftop/panels/NamespacePanel.java [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+      Snapshot snapshot = root.getLag().getSnapshot();
+      addFirstRow(root, factor, nodes, snapshot);
+      addNodes(root, factor, nodes, snapshot, selectedLabel);
+    }
+  }
+
+  /**
+   * Add first row (to go up a folder).
+   */
+  protected abstract void addFirstRow(Node root, double factor, Collection<Node> nodes, Snapshot snapshot);
```
Comment (panghy) [inline src/main/java/com/wavefront/tools/wftop/panels/NamespacePanel.java]: group all abstract methods together

File: src/main/java/com/wavefront/tools/wftop/panels/NamespacePanel.java [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+
+  /**
+   * Now sort and add the nodes for this folder.
+   */
+  protected abstract void addNodes(Node root, double factor, Collection<Node> nodes, Snapshot snapshot, String selectedLabel);
+
+
```
Comment (panghy) [inline src/main/java/com/wavefront/tools/wftop/panels/NamespacePanel.java]: blank line

**Comment by ktjoanna** [comment]
Added Export Feature in addition to these changes #11

**Pull request closed without merging**